Accessibly disable save button on edit question screen

chrispymm opened the pull request:
Updates the question edit view save button to use `aria-disabled` instead of `disabled`.

This allows users of assistive technology to access the button in its disabled state.

Becasue `aria-disabled` only updates the buttons semantics, we add in liseners to capture `click`, `submit` and `keydown` events to prevent the submission of the form when the button is disabled. 

We also override these listeners for actions that need to submit the form: Add New Component and Add Content

stevenburnell-moj commented:
This seems like a lot of effort with the result of quite complicated/confusing code. 

If I'm understanding the problem correctly, you're basically trying to apply aria-disabled to the button instead of using the native 'disabled' attributes, to help avoid some issues with accessibility handling. There might be more, but it's not obvious (maybe due to GitHub comparison screens).

Either way, strikes me that you already had something like this:

```
saveRequired(required) {
      if(required) { 
         this.$form.find(":submit").prop("value", this.text.actions.save );
         this.$form.find(":submit").prop("disabled", false);
     }
     else {
        this.$form.find(":submit").prop("value", this.text.actions.saved );
        this.$form.find(":submit").prop("disabled", true);
     }
 }
```

Which, presumably, gave the functionality required albeit using the disable attribute to prevent form submit.

Would it not have been simply easier to repurpose that function to disable the form submit and apply whatever (aria-disabled) additions to the button?


e.g. Wouldn't this small change do pretty much everything you're doing now with a boat-load confusing event handling?
```
saveRequired(required) {
      var $submit = this.$form.find(":submit");
      if(required) { 
         $submit.prop("value", this.text.actions.save );
         $submit.prop("aria-disabled", false); // assume removes it
     }
     else {
        $submit.prop("value", this.text.actions.saved );
        $submit.prop("aria-disabled", true);
     }
 }
```

// Add something like this to form submit() event via best/easiest solution
```
this.$form.on("submit", function(e ) {
   if( $(this).find(":submit").prop("aria-disabled") ) {
      // prevent submit
   }
});
```

chrispymm commented:
I'm struggling to remember now, but I seem to think that there were reasons why it wasn't this simple.... I'll take a nother look and see if I can simplify it. IIRC the extra complexity was around the fact that it was requiring extra work to prevent the keyboard events from triggering the submission.  I know at one point I was also trying to stop the Voicoever submit keyboard command (Cmd-Opt-Space) to no avail whatsoever... so it could be that the complexity was introduced by that, and I can simplify more.

I'll revisit and see if I can simplify.

chrispymm commented:
So I've updated this to simplify: moved the click  and submit handler back to being inline, and the keyboard listener was no longer required.

I've left in the submitEnabled class property as it needs to be queried from elsewhere as adding a component or a content area needs to be able to read and override it in order to submit the form to add the new thing.

stevenburnell-moj commented on app/javascript/src/controller_pages.js:
```diff
@@ -151,18 +151,29 @@ PagesController.create = function() {
 class DataController {
```
Not recommending a change here but...

Originally, `DataController` was intended as some (generic??) component that could handle storing data on client side, updating if necessary, and then sending off to backend for permanent storage. I dare say that never happened as it was influenced by earlier expectation things might move toward frontend-app like control (which we were not keen to do for reasons of available developer resource). 

Anyway, looking through things now it definitely looks like the `DataController` name is not very descriptive as, if anything, its seems to be more of a `FormSubmitController` or `PageActionController` (or ??); whatever is appropriate and allows for reuse (I guess, if even possible). 

Doesn't matter much but if we have a current opportunity, given that I'd expect you now have a good understanding of exactly what functionality it provides, perhaps it could be renamed to be more descriptive and help with future development, etc.

chrispymm replied:
Yep, you're totally right, and I will likely be doing this in my next ticket.   I need to add (some) of the submit button functionality to forms in the settings screen, so this code is likely to be revisited then, and so I'll look at a rename then.

chrispymm commented on app/javascript/src/controller_pages.js:
```diff
@@ -151,18 +151,29 @@ PagesController.create = function() {
-    $form.on("submit", () => { 
-      this.update();
-      this.removeBeforeUnloadListener();
+    this.$form.find(':submit').on('click', (event) => {
```
👍

chrispymm commented on app/javascript/src/controller_pages.js:
```diff
@@ -151,18 +151,29 @@ PagesController.create = function() {
-      this.removeBeforeUnloadListener();
+    this.$form.find(':submit').on('click', (event) => {
+      if(this.submitEnabled) {
+        window.removeEventListener('beforeunload', this.beforeUnloadListener, {capture: true});
```
Good spot! Fixed.
